### PR TITLE
update to fmt 10.1.1, fast_double_parser 0.7.0

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #3763.
Contributes to #5691.

Proposes updated the projects vendored in here as git submodules to their latest versions:

* `boostorg/compute`: no change (already on [v1.83.0, September 2021](https://github.com/boostorg/compute/tags))
* `fmtlib/fmt`: from `v8.1.1` ([January 2022](https://github.com/fmtlib/fmt/releases/tag/8.1.1)) to `v10.1.1` ([August 2023](https://github.com/fmtlib/fmt/releases/tag/10.1.1))
* `lemire/fast_double_parser`: from `v0.3.0` ([November 2020](https://github.com/lemire/fast_double_parser/releases/tag/v0.3.0)) to `v0.7.0` ([December 2022](https://github.com/lemire/fast_double_parser/releases/tag/v0.7.0))
* `libeigen/eigen`: no change (already on [v3.4.0, August 2021](https://gitlab.com/libeigen/eigen/-/tags))

### How I did that

```shell
# fast_double_parser
cd ./external_libs/fast_double_parser
git fetch origin v0.7.0
git checkout v0.7.0
cd ../../
git add ./external_libs/fast_double_parser

# fmt
cd ./external_libs/fmt
git fetch origin 10.1.1
git checkout 10.1.1
cd ../../
git add ./external_libs/fmt
```

### Benefits of this change

Pulls in bugfixes, security patches, performance improvements.

I'm hoping it'll help with R 4.3 support via changes like this:

* https://github.com/fmtlib/fmt/issues/2898
* https://github.com/fmtlib/fmt/issues/3533

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -324,7 +324,7 @@ endif()
 if(UNIX OR MINGW OR CYGWIN)
   set(
     CMAKE_CXX_FLAGS
-    "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas -Wno-return-type"
+    "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas -Wno-return-type -Wno-stringop-overflow"
```
Comment (jameslamb) [inline CMakeLists.txt:327]: This is used to suppress the following warning that shows up only with MinGW (gcc 10):

```text
[ 60%] Building CXX object CMakeFiles/lightgbm_objs.dir/src/network/linkers_mpi.cpp.obj

d:\a\lightgbm\lightgbm\lightgbm.rcheck\00_pkg_src\lightgbm\src\external_libs\fmt\include\fmt\format.h: In function 'void fmt::v10::detail::format_hexfloat(Float, int, fmt::v10::detail::float_specs, fmt::v10::detail::buffer<char>&) [with Float = double; typename std::enable_if<(! std::integral_constant<bool, (std::numeric_limits<_Tp>::digits == 106)>::value), int>::type <anonymous> = 0]':
d:\a\lightgbm\lightgbm\lightgbm.rcheck\00_pkg_src\lightgbm\src\external_libs\fmt\include\fmt\format.h:1370:8: note: at offset -2 to object 'buffer' with size 10 declared here
 1370 |   Char buffer[digits10<UInt>() + 1] = {};
      |        ^~~~~~
```

([example build link](https://github.com/microsoft/LightGBM/actions/runs/6067151932/job/16458483403?pr=6074))

Per https://github.com/fmtlib/fmt/issues/1810#issuecomment-671568943, this was a false positive and a bug in `gcc` 10 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95353). Seems that maybe it wasn't backported in MinGW as well?

File: CMakeLists.txt [hunk 2]
```diff
@@ -326,6 +326,13 @@ if(UNIX OR MINGW OR CYGWIN)
     CMAKE_CXX_FLAGS
     "${CMAKE_CXX_FLAGS} -std=c++11 -pthread -Wextra -Wall -Wno-ignored-attributes -Wno-unknown-pragmas -Wno-return-type"
   )
+  if(MINGW)
+    # ignore this warning: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95353
+    set(
+      CMAKE_CXX_FLAGS
+      "${CMAKE_CXX_FLAGS} -Wno-stringop-overflow"
+    )
+  endif()
```
Comment (jameslamb) [inline CMakeLists.txt:335]: This is used to suppress the following warning that shows up only with MinGW (gcc 10):

```text
[ 60%] Building CXX object CMakeFiles/lightgbm_objs.dir/src/network/linkers_mpi.cpp.obj

d:\a\lightgbm\lightgbm\lightgbm.rcheck\00_pkg_src\lightgbm\src\external_libs\fmt\include\fmt\format.h: In function 'void fmt::v10::detail::format_hexfloat(Float, int, fmt::v10::detail::float_specs, fmt::v10::detail::buffer<char>&) [with Float = double; typename std::enable_if<(! std::integral_constant<bool, (std::numeric_limits<_Tp>::digits == 106)>::value), int>::type <anonymous> = 0]':
d:\a\lightgbm\lightgbm\lightgbm.rcheck\00_pkg_src\lightgbm\src\external_libs\fmt\include\fmt\format.h:1370:8: note: at offset -2 to object 'buffer' with size 10 declared here
 1370 |   Char buffer[digits10<UInt>() + 1] = {};
      |        ^~~~~~
```

([example build link](https://github.com/microsoft/LightGBM/actions/runs/6067151932/job/16458483403?pr=6074))

Per https://github.com/fmtlib/fmt/issues/1810#issuecomment-671568943, this was a false positive and a bug in `gcc` 10 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95353). Seems that maybe it wasn't backported in MinGW as well?

**Comment by jameslamb** [comment]
@guolinke @shiyu1994 could I get help with a review on this?

In addition to the bugfixes and performance improvements in these new versions of these libraries, I think this PR will help with fully testing the R package against R 4.3 and with efforts like #5981 to make the project compatible with C++20.

**Review by guolinke — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks so much @guolinke !

**Pull request merged**